<a href="https://colab.research.google.com/github/kef135lib/advPy4lib/blob/main/KFloyd_Week_1_Homework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 1 Homework

(Be sure to save this to Drive before starting)

For this week we are going to practice requesting information from an API and manipulating it as JSON and as a Dataframe.

We are going to keep working with Collection Builder. Head over to the [Collection Builder Example Projects](https://collectionbuilder.github.io/cb-examples/browse.html) listing and find a site that interests you. Spend some time browsing the website to get acquainted with its contents before proceeding with the questions.

In [ ]:
#The Libraries we will need
import requests
import json
import pandas as pd
from random import randrange

In [ ]:
# @markdown **H1** Find the URL of for the  _metadata.json_ of your selected site and paste it in here.
json_cb_url = "https://cpplibrary.github.io/postcards-compound/data/metadata.json" # @param {"type":"string","placeholder":""}

response_cb = requests.get(json_cb_url)
#all of the data from the sites
total_cb_data = response_cb.json()
#just the 'objects' indexed in the site
objects_cb_data = total_cb_data['objects']
print("Data fetched and JSON constructed!")


Data fetched and JSON constructed!


**H2** Run the 'show random cell' a few times to get a sense of what kind of data is in your chosen site.

In [ ]:
#Run this a few times to see random entries in the JSON data that you retrieved.
random_entry = randrange(len(objects_cb_data))
print(json.dumps(objects_cb_data[random_entry],indent=2))

{
  "objectid": "coll008",
  "filename": "cpp_c0015_b0001_i007a.jpg",
  "compound": "cpp_c0015_b0001_i007b.jpg",
  "title": "Antez",
  "description": "[1921-1944 \"Antez\", a beautiful chestnut stallion with flaxen mane and tail, is one of a stable of more than 40 full blooded Arabian horses owned by the W.K. Kellogg Arabian Horse Ranch, of Pomona, California. The owner of this ranch is the well known cereal manufacturer of Battle Creek, Michigan. \"Antez\" was the Champion Arabian Stallion at the Los Angeles County Fair in 1928.]",
  "postcard_type": "Image;StillImage",
  "subject": "Arabian horse; California State Polytechnic University, Pomona; horse; postcards; W.K. Kellogg Arabian Horse Center",
  "location": "Pomona ",
  "state": "California",
  "publisher": "California State Polytechnic University, Pomona",
  "publisher_location": "Pomona, CA",
  "source": "W.K. Kellogg Arabian Horse Library Postcard and Notecard Collection",
  "identifier": "cpp_c0015_b0001_i007a",
  "rights": 

## Different Fields!

You can create any number of datafields and correpsonding values for those fields with a Collection Builder site. So the results from the output from the cell above will probably have wildly different results from [my site](https://elibtronic.github.io/AIL_Database/data.html) that we worked on in the Workalong notebook.


You can see all of these options from grabbing the _facets.json_ URL from a collection builder homepage. Find that URL for your selected site for **H3**

In [ ]:
# @markdown **H3** Find the URL of the facets.json for your selected site and paste it in here.
json_cb_facets_url = "https://cpplibrary.github.io/postcards-compound/data/facets.json" # @param {"type":"string","placeholder":""}

response_cb = requests.get(json_cb_facets_url)
#all of the data from the sites
total_facet_data = response_cb.json()
#just the 'objects' indexed in the site
print("Data fetched and JSON constructed!")


Data fetched and JSON constructed!


Run the cell below to see what the facets used for this site:

In [ ]:
print(json.dumps(total_facet_data,indent=2))

{
  "facets": [
    {
      "field": "postcard_type",
      "terms": {
        "image": 128,
        "stillimage": 128
      }
    },
    {
      "field": "photographer",
      "terms": {}
    },
    {
      "field": "subject",
      "terms": {
        "arabian horse": 103,
        "arabian horses": 1,
        "architecture": 6,
        "boats": 1,
        "california state polytechnic university, pomona": 79,
        "campuses": 2,
        "cereal products industry": 5,
        "factories": 3,
        "highways": 1,
        "horse": 104,
        "horse show": 1,
        "horse shows": 4,
        "horsemanship": 37,
        "horses": 1,
        "industrial recreation": 2,
        "kellogg company": 5,
        "kellogg house, pomona": 3,
        "kellogg, will keith, 1860-1951": 1,
        "macatawa, lake (mich.)": 1,
        "postcards": 128,
        "show horses": 2,
        "stables": 7,
        "university campuses": 3,
        "w.k. kellogg arabian horse center": 96
      }
    },


**H3** Try to read the JSON that is displayed to the screen. What possible facets are there in the the data. Fill them in the markdown list in the cell below. Add as many entries as you need to describe the site

Facets in my site are:
- postcard_type
- photographer
- subject
- location
- state
- publisher
- publisher_location
- original_format
- format

**H4** Which facet shows up the most for your selected site? Put your answer in the cell below


This facet value that shows up the most is "arabian horse." It is used 103 times on my selected site.

**H5** Turn the site _metadata.json_ data into a dataframe by completing the cell below. Part of it is written for you. You just need to modify line 4.

In [ ]:
#H5

cb_data = pd.read_json(json_cb_url)
cb_normalized = pd.json_normalize(cb_data['objects'])
cb_normalized

,objectid,filename,compound,title,description,postcard_type,subject,location,state,publisher,...,identifier,rights,standardized_rights,type,format,language,latitude,longitude,cdmid,reference_url
0,coll002,cpp_c0015_b0001_i001a.jpg,cpp_c0015_b0001_i001b.jpg,Abu Farwa,Verso [1940-1972 Abu Farwa's name is practical...,Image;StillImage,horses; postcards,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i001a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i001a,https://cpplibrary.github.io/postcards-compoun...
1,coll003,cpp_c0015_b0001_i002a.jpg,cpp_c0015_b0001_i002b.jpg,Alyf,"Verso: [""Alyf"" foaled in 1938, was the winner ...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i002a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i002a,https://cpplibrary.github.io/postcards-compoun...
2,coll004,cpp_c0015_b0001_i003a.jpg,cpp_c0015_b0001_i003b.jpg,Alyf,"Recto: [Arabian Stallion, Alyf, Pomona Quarter...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i003a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i003a,https://cpplibrary.github.io/postcards-compoun...
3,coll005,cpp_c0015_b0001_i004a.jpg,cpp_c0015_b0001_i004b.jpg,Alyf,"Verso: [Alyf, foaled 1938. Head study of eight...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i004a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i004a,https://cpplibrary.github.io/postcards-compoun...
4,coll006,cpp_c0015_b0001_i005a.jpg,cpp_c0015_b0001_i005b.jpg,Alyf,"Verso: [Alyf, foaled 1938. Head study of eight...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i005a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i005a,https://cpplibrary.github.io/postcards-compoun...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
123,coll125,cpp_c0015_b0001_i124a.jpg,cpp_c0015_b0001_i124b.jpg,"Stables on W.K. Kellogg Arabian Horse Ranch, P...","Verso: [The W.K. Kellogg Arabian Horse Ranch, ...",Image;StillImage,"California State Polytechnic University, Pomon...",Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i124a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i124a,https://cpplibrary.github.io/postcards-compoun...
124,coll126,cpp_c0015_b0001_i125a.jpg,cpp_c0015_b0001_i125b.jpg,"Stables on W.K. Kellogg Arabian Horse Ranch, P...","Verso: [The W.K. Kellogg Arabian Horse Ranch, ...",Image;StillImage,"California State Polytechnic University, Pomon...",Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i125a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i125a,https://cpplibrary.github.io/postcards-compoun...
125,coll127,cpp_c0015_b0001_i126a.jpg,cpp_c0015_b0001_i126b.jpg,Union Plaza,"Verso: [Union Plaza, the original stables on t...",Image;StillImage,Architecture; California State Polytechnic Uni...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i126a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i

In [ ]:
#Run this a few times to see random entries in your dataframe
cb_normalized.sample(10)

,objectid,filename,compound,title,description,postcard_type,subject,location,state,publisher,...,identifier,rights,standardized_rights,type,format,language,latitude,longitude,cdmid,reference_url
107,coll109,cpp_c0015_b0001_i108a.jpg,cpp_c0015_b0001_i108b.jpg,"Macatawa Yacht Harbor, Holland, Mich.","Recto: [Macatawa Yacht Harbor, Holland, Mich.]",Image;StillImage,"Boats; Macatawa, Lake (Mich.); postcards",Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i108a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i108a,https://cpplibrary.github.io/postcards-compoun...
93,coll095,cpp_c0015_b0001_i094a.jpg,cpp_c0015_b0001_i094b.jpg,The Three Graces,"Recto: [The Three Graces, Arabian mares and co...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i094a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i094a,https://cpplibrary.github.io/postcards-compoun...
8,coll010,cpp_c0015_b0001_i009a.jpg,cpp_c0015_b0001_i009b.jpg,Antez,"[1921-1944 ""Antez"", a beautiful chestnut stall...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i009a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i009a,https://cpplibrary.github.io/postcards-compoun...
59,coll061,cpp_c0015_b0001_i060a.jpg,cpp_c0015_b0001_i060b.jpg,Maazal,"Maazal, foaled April 25, 1953. Verso: [Three Y...",Image;StillImage,Arabian horse; horse; postcards,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i060a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i060a,https://cpplibrary.github.io/postcards-compoun...
85,coll087,cpp_c0015_b0001_i086a.jpg,cpp_c0015_b0001_i086b.jpg,Six-Horse Percheron Hitch,"Verso: [California State Polytechnic College, ...",Image;StillImage,Arabian horse; California State Polytechnic Un...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i086a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i086a,https://cpplibrary.github.io/postcards-compoun...
119,coll121,cpp_c0015_b0001_i120a.jpg,cpp_c0015_b0001_i120b.jpg,"Stables on W.K. Kellogg Arabian Horse Ranch, P...","Verso: [The W.K. Kellogg Arabian Horse Ranch, ...",Image;StillImage,"California State Polytechnic University, Pomon...",Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i120a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i120a,https://cpplibrary.github.io/postcards-compoun...
114,coll116,cpp_c0015_b0001_i115a.jpg,cpp_c0015_b0001_i115b.jpg,Cal Poly Pomona,Recto: [Cal Poly Pomona],Image;StillImage,Architecture; California State Polytechnic Uni...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i115a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i115a,https://cpplibrary.github.io/postcards-compoun...
83,coll085,cpp_c0015_b0001_i084a.jpg,cpp_c0015_b0001_i084b.jpg,Sikin,"Recto: [Sikin, foaled 1933. W.K. Kellogg's Ins...",Image;StillImage,Arabian horse; horse; postcards; W.K. Kellogg ...,Pomona,California,"California State Polytechnic University, Pomona",...,cpp_c0015_b0001_i084a,Public domain,http://rightsstatements.org/vocab/NoC-US/1.0/,Image;StillImage,image/jpeg,en,34.0589,-117.8194,cpp_c0015_b0001_i084a,https://cpplibrary.github.io/postcards-compoun...
7,coll009,cpp_c0015_b0001_i0

**H6** This is for bonus!
Try to write a pandas statement that will display your answer from **H4**. There are plenty of ways to do this, so feel free to use your imagination. What I have started for you here might not work, so feel free to change it as needed.

In [ ]:
number = cb_normalized["subject"].str.contains("Arabian horse;").sum()
print("This facet value that shows up the most is \"arabian horse.\" It is used ",number, " times on my selected site.")

This facet value that shows up the most is "arabian horse." It is used  103  times on my selected site.



## All Done!

Congrats you are finished with this week's work. First save as PDF so that you can upload to the Course site then click the `Share` button above and enter `libraryjuicepresspython@gmail.com` so I can get a shared copy of your work.